File: tests/kotlin/test_2_1_2_graph_algorithms.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.collections.shouldContain
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.apache.tinkerpop.gremlin.tinkercat.algorithms.breadthFirstSearch
import org.apache.tinkerpop.gremlin.tinkercat.algorithms.depthFirstSearch
import org.apache.tinkerpop.gremlin.tinkercat.algorithms.hasCycle
import org.apache.tinkerpop.gremlin.tinkercat.algorithms.isConnected
import org.apache.tinkerpop.gremlin.tinkercat.algorithms.shortestPath
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 2.1.2: Graph Algorithms.
 *
 * Validates BFS, DFS, shortest path, connectivity, and cycle detection
 * implemented as extension functions on TinkerCat.
 * Mirrors tests/python/test_2_1_2_graph_algorithms.py.
 */
class Test_2_1_2_GraphAlgorithms : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "bfs on single vertex returns just that vertex" {
        val v = graph.addVertex(TinkerVertex.T.label, "node")
        val result = graph.breadthFirstSearch(v).toList()
        result shouldHaveSize 1
        result[0].id() shouldBe v.id()
    }

    "bfs visits all vertices in a linear chain" {
        val v1 = graph.addVertex(TinkerVertex.T.label, "node")
        val v2 = graph.addVertex(TinkerVertex.T.label, "node")
        val v3 = graph.addVertex(TinkerVertex.T.label, "node")
        v1.addEdge("next", v2)
        v2.addEdge("next", v3)
        val result = graph.breadthFirstSearch(v1).toList()
        result shouldHaveSize 3
    }

    "bfs visits closer vertices before farther ones" {
        val root = graph.addVertex(TinkerVertex.T.label, "node")
        val child1 = graph.addVertex(TinkerVertex.T.label, "node")
        val child2 = graph.addVertex(TinkerVertex.T.label, "node")
        val grandchild = graph.addVertex(TinkerVertex.T.label, "node")
        root.addEdge("link", child1)
        root.addEdge("link", child2)
        child1.addEdge("link", grandchild)
        val result = graph.breadthFirstSearch(root).toList()
        val rootIdx = result.indexOfFirst { it.id() == root.id() }
        val grandIdx = result.indexOfFirst { it.id() == grandchild.id() }
        rootIdx shouldBe 0
        (grandIdx > 2) shouldBe true
    }

    "dfs on single vertex returns just that vertex" {
        val v = graph.addVertex(TinkerVertex.T.label, "node")
        val result = graph.depthFirstSearch(v).toList()
        result shouldHaveSize 1
    }

    "dfs visits all vertices in a chain" {
        val v1 = graph.addVertex(TinkerVertex.T.label, "node")
        val v2 = graph.addVertex(TinkerVertex.T.label, "node")
        val v3 = graph.addVertex(TinkerVertex.T.label, "node")
        v1.addEdge("next", v2)
        v2.addEdge("next", v3)
        val result = graph.depthFirstSearch(v1).toList()
        result shouldHaveSize 3
    }

    "bfs and dfs visit same set of vertices" {
        val v1 = graph.addVertex(TinkerVertex.T.label, "n")
        val v2 = graph.addVertex(TinkerVertex.T.label, "n")
        val v3 = graph.addVertex(TinkerVertex.T.label, "n")
        v1.addEdge("e", v2)
        v2.addEdge("e", v3)
        val bfsIds = graph.breadthFirstSearch(v1).map { it.id() }.toSet()
        val dfsIds = graph.depthFirstSearch(v1).map { it.id() }.toSet()
        bfsIds shouldBe dfsIds
    }

    "shortest path between adjacent vertices has length 2" {
        val a = graph.addVertex(TinkerVertex.T.label, "node")
        val b = graph.addVertex(TinkerVertex.T.label, "node")
        a.addEdge("link", b)
        val path = graph.shortestPath(a, b)
        path shouldNotBe null
        path!! shouldHaveSize 2
    }

    "shortest path returns null for disconnected vertices" {
        val a = graph.addVertex(TinkerVertex.T.label, "node")
        val b = graph.addVertex(TinkerVertex.T.label, "node")
        graph.shortestPath(a, b) shouldBe null
    }

    "shortest path from vertex to itself is single-element list" {
        val a = graph.addVertex(TinkerVertex.T.label, "node")
        val path = graph.shortestPath(a, a)
        path shouldNotBe null
        path!! shouldHaveSize 1
    }

    "connected graph is detected as connected" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val c = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("e", b)
        b.addEdge("e", c)
        graph.isConnected() shouldBe true
    }

    "disconnected graph is detected as disconnected" {
        graph.addVertex(TinkerVertex.T.label, "n")
        graph.addVertex(TinkerVertex.T.label, "n")
        graph.isConnected() shouldBe false
    }

    "acyclic graph has no cycle" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val c = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("e", b)
        b.addEdge("e", c)
        graph.hasCycle() shouldBe false
    }

    "graph with cycle is detected" {
        val a = graph.addVertex(TinkerVertex.T.label, "n")
        val b = graph.addVertex(TinkerVertex.T.label, "n")
        val c = graph.addVertex(TinkerVertex.T.label, "n")
        a.addEdge("e", b)
        b.addEdge("e", c)
        c.addEdge("e", a)
        graph.hasCycle() shouldBe true
    }
})

File: tests/npm/test_2_1_2_graph_algorithms.ts
/**
 * Tests for Task 2.1.2: Graph Algorithms.
 * Mirrors tests/python/test_2_1_2_graph_algorithms.py
 */
import { describe, it, expect, beforeEach } from 'vitest'
import tc from './tc'
import type { TinkerCat, Vertex } from './tinkercat.d.ts'
import { buildModernGraph } from './helpers'

describe('Graph Algorithms', () => {
  let graph: TinkerCat

  beforeEach(() => { graph = tc.createTinkerCat() })

  it('BFS visits all reachable vertices', () => {
    const { marko } = buildModernGraph(graph)
    const visited: Vertex[] = tc.breadthFirstSearch(graph, marko)
    expect(visited.length).toBeGreaterThanOrEqual(1)
  })

  it('DFS visits all reachable vertices', () => {
    const { marko } = buildModernGraph(graph)
    const visited: Vertex[] = tc.depthFirstSearch(graph, marko)
    expect(visited.length).toBeGreaterThanOrEqual(1)
  })

  it('BFS includes the start vertex', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const result: Vertex[] = tc.breadthFirstSearch(graph, a)
    expect(result.some((v: Vertex) => tc.getPropertyValue(v, 'name') === 'A')).toBe(true)
  })

  it('DFS includes the start vertex', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const result: Vertex[] = tc.depthFirstSearch(graph, a)
    expect(result.some((v: Vertex) => tc.getPropertyValue(v, 'name') === 'A')).toBe(true)
  })

  it('shortestPath returns a path between connected vertices', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const c: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'C' })
    tc.addEdge(a, 'link', b, {})
    tc.addEdge(b, 'link', c, {})
    const path: Vertex[] | null = tc.shortestPath(graph, a, c)
    expect(path).not.toBeNull()
    expect(path!.length).toBeGreaterThan(0)
  })

  it('shortestPath returns null for disconnected vertices', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const path: Vertex[] | null = tc.shortestPath(graph, a, b)
    expect(path).toBeNull()
  })

  it('shortestPath self-loop returns single vertex', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const path: Vertex[] | null = tc.shortestPath(graph, a, a)
    expect(path).not.toBeNull()
    expect(path!.length).toBe(1)
  })

  it('hasCycle returns false for acyclic graph', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    tc.addEdge(a, 'link', b, {})
    expect(tc.hasCycle(graph)).toBe(false)
  })

  it('hasCycle returns true for cyclic graph', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const c: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'C' })
    tc.addEdge(a, 'link', b, {})
    tc.addEdge(b, 'link', c, {})
    tc.addEdge(c, 'link', a, {})
    expect(tc.hasCycle(graph)).toBe(true)
  })

  it('findConnectedComponents returns one component for connected graph', () => {
    const a: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    const b: Vertex = tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    tc.addEdge(a, 'link', b, {})
    const components: Vertex[][] = tc.findConnectedComponents(graph)
    expect(components.length).toBe(1)
  })

  it('findConnectedComponents returns multiple components for disconnected graph', () => {
    tc.addVertexWithLabel(graph, 'node', { name: 'A' })
    tc.addVertexWithLabel(graph, 'node', { name: 'B' })
    const components: Vertex[][] = tc.findConnectedComponents(graph)
    expect(components.length).toBe(2)
  })

  it('empty graph has no cycle', () => {
    expect(tc.hasCycle(graph)).toBe(false)
  })

  it('getGraphStatistics returns vertex and edge counts', () => {
    buildModernGraph(graph)
    const stats = tc.getGraphStatistics(graph)
    expect(stats.vertexCount).toBe(6)
    expect(stats.edgeCount).toBe(6)
  })
})
